### PR TITLE
th12

### PR DESCRIPTION
**Pull request by raviddog**

### CODE REVIEW

File: project/thscoreboard/replays/game_fields.py [hunk 1]
```diff
@@ -156,7 +179,20 @@ def GetFormatPower(game_id: str, power: Optional[int]) -> str:
     'th08': None,
     'th09': None,
     'th10': None,
-    'th11': 5
+    'th11': 5,
+    'th12': 4
+})
+
+_bomb_pieces = immutabledict({
```
Comment (n-rook) [inline project/thscoreboard/replays/game_fields.py:186]: Add a comment saying what this dict is

File: project/thscoreboard/replays/game_fields.py [hunk 2]
```diff
@@ -156,7 +179,20 @@ def GetFormatPower(game_id: str, power: Optional[int]) -> str:
...
+    'th01': None,
+    'th05': None,
```
Comment (n-rook) [inline project/thscoreboard/replays/game_fields.py:188]: We support th02-04 too, so we should specify them as well. (Or, alternatively, I think I'd be OK with only including the games with bomb pieces in them here, and calling `_bomb_pieces.get(game_id, None)` instead.)

File: project/thscoreboard/replays/replay_parsing.py [hunk 3]
```diff
@@ -530,6 +532,76 @@ def _Parse11(rep_raw):
...
+    rep_stages = []
+    for stage in replay.stages:
+        s = ReplayStage()
+        s.stage = stage.stage
+        s.score = stage.score * 10
+        s.power = stage.power
+        s.piv = (math.trunc(stage.piv / 1000)) * 10
+        s.lives = stage.lives
+        s.life_pieces = stage.life_pieces
+        #   fix zun fuckery
```
Comment (n-rook) [inline project/thscoreboard/replays/replay_parsing.py:554]: lol

File: project/thscoreboard/replays/replay_parsing.py [hunk 4]
```diff
@@ -530,6 +532,76 @@ def _Parse11(rep_raw):
...
+            stage.piv = None
+            stage.lives = None
+            stage.life_pieces = None
+            stage.bombs = None
+            stage.bomb_pieces = None
+            stage.graze = None
+
+    r_type = game_ids.ReplayTypes.REGULAR
+    if len(rep_stages) == 1 and replay.header.difficulty != 4:
+        r_type = game_ids.ReplayTypes.STAGE_PRACTICE
```
Comment (n-rook) [inline project/thscoreboard/replays/replay_parsing.py:588]: I think if we're going to detect stage practice, we need a test that's stage practice, just to check.

**Review by n-rook — CHANGES_REQUESTED** [review]
Looks basically good, just a few minor comments

File: project/thscoreboard/replays/game_fields.py (hunk 2 above)
Comment (raviddog) [reply project/thscoreboard/replays/game_fields.py:188]: games that don't include stage splits shouldn't ever be called in here, though theres no harm in it so i guess ill put em all in

File: project/thscoreboard/replays/replay_parsing.py (hunk 4 above)
Comment (raviddog) [reply project/thscoreboard/replays/replay_parsing.py:588]: theres already one, test_replay_parsing.py line 195

**Comment by raviddog** [comment]
wait why do these tests fail it works on my local branch

**Comment by raviddog** [comment]
the more i fiddle with this the more it breaks

File: project/thscoreboard/replays/replay_parsing.py (hunk 4 above)
Comment (n-rook) [reply project/thscoreboard/replays/replay_parsing.py:601]: I mean a test that actually tests this code in particular; that is, a test for UFO stage practice

File: project/thscoreboard/replays/test_replay_parsing.py [hunk 5]
```diff
@@ -19,7 +20,7 @@ def testHard1cc(self):
         self.assertEqual(r.shot, 'ReimuA')
         self.assertEqual(r.score, 92245410)
         self.assertEqual(r.name, 'nrook   ')
-        self.assertEqual(r.replay_type, 1)
+        self.assertEqual(r.replay_type, game_ids.ReplayTypes.REGULAR)
```
Comment (n-rook) [inline project/thscoreboard/replays/test_replay_parsing.py:23]: this is way easier to read btw, thanks for doing this

**Review by n-rook — CHANGES_REQUESTED** [review]

**Comment by n-rook** [comment]
python datetimes can be really crazy and even half of professional developers don't know how to handle them right. basically the best thing to do is to make everything an "aware datetime" by passing in a timezone immediately, because otherwise some library operations will assume the datetime represents local time. this tends to manifest as tests failing on github but passing locally (or the other way around) because the test env on github is UTC but your local pc probably isn't. fortunately it seems to be fine now

**Review by n-rook — APPROVED** [review]

**Pull request merged**